modules/library/src/main/scala/zio/elasticsearch/query/ElasticIntervalQuery.scala

```diff
@@ -0,0 +1,356 @@
+package zio.elasticsearch.query
+
+import zio.elasticsearch.ElasticPrimitive.ElasticPrimitiveOps
+import zio.elasticsearch.query.options.{HasAnalyzer, HasUseField}
+import zio.json.ast.Json
+import zio.json.ast.Json.{Arr, Obj, Str}
+import zio.{Chunk, NonEmptyChunk}
+import zio.elasticsearch.Field
+
+sealed trait BoundType
+sealed trait Inclusive extends BoundType
+sealed trait Exclusive extends BoundType
+
+case object InclusiveBound extends Inclusive
+case object ExclusiveBound extends Exclusive
+
+final case class Bound[B <: BoundType](value: String, boundType: B)
+
+sealed trait IntervalRule {
+  private[elasticsearch] def toJson: Json
+}
+
+final case class GreaterThanInterval(value: String) extends IntervalRule {
+  private[elasticsearch] def toJson: Json = Str(value)
+}
+
+final case class GreaterThanOrEqualToInterval(value: String) extends IntervalRule {
+  private[elasticsearch] def toJson: Json = Str(value)
+}
+
+final case class LessThanInterval(value: String) extends IntervalRule {
+  private[elasticsearch] def toJson: Json = Str(value)
+}
+
+final case class LessThanOrEqualToInterval(value: String) extends IntervalRule {
+  private[elasticsearch] def toJson: Json = Str(value)
+}
+
+private[elasticsearch] final case class IntervalAllOf[S](
+  intervals: Chunk[IntervalRule],
+  maxGaps: Option[Int],
+  ordered: Option[Boolean],
+  filter: Option[IntervalFilter[S]]
+) extends IntervalRule {
+
+  def filter(f: IntervalFilter[S]): IntervalAllOf[S] = copy(filter = Some(f))
+  def maxGaps(g: Int): IntervalAllOf[S]              = copy(maxGaps = Some(g))
+  def orderedOn: IntervalAllOf[S]                    = copy(ordered = Some(true))
+  def orderedOff: IntervalAllOf[S]                   = copy(ordered = Some(false))
+  private[elasticsearch] def toJson: Json            =
```

Comment:
Check do we need empty lines between methods. Also, I would definitely add an empty line before `toJson`.